File: app/src/main/java/org/cuieney/videolife/ui/fragment/video/VideoDetailFragment.java
package org.cuieney.videolife.ui.fragment.video;

import android.content.res.ColorStateList;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.graphics.Palette;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SimpleTarget;
import com.konifar.fab_transformation.FabTransformation;

import org.cuieney.videolife.R;
import org.cuieney.videolife.common.image.ImageLoader;
import org.cuieney.videolife.common.utils.DateUtil;
import org.cuieney.videolife.common.utils.LogUtil;
import org.cuieney.videolife.entity.kaiyanBean.DataBean;
import org.cuieney.videolife.common.base.BaseBackFragment;
import org.cuieney.videolife.ui.video.JumpUtils;


/**
 * Created by cuieney on 17/2/25.
 */

public class VideoDetailFragment extends BaseBackFragment {

    ImageView mImgDetail;
    Toolbar mToolbar;
    ImageView bgImage;
    TextView title;
    TextView type;
    TextView description;
    FloatingActionButton mFab;


    private DataBean dataBean;
    private CollapsingToolbarLayout collToolBar;


    public static VideoDetailFragment newInstance(DataBean dataBean) {

        Bundle args = new Bundle();
        args.putParcelable(ARG_ITEM, dataBean);
        VideoDetailFragment fragment = new VideoDetailFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        dataBean = getArguments().getParcelable(ARG_ITEM);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.video_home_detail_fragment, container, false);
        initView(view);
        return view;
    }

    private void initView(View view) {
        mToolbar = ((Toolbar) view.findViewById(R.id.toolbar));
        collToolBar = ((CollapsingToolbarLayout) view.findViewById(R.id.toolbar_layout));
        mImgDetail = (ImageView) view.findViewById(R.id.img_detail);
        bgImage = (ImageView) view.findViewById(R.id.bg_image);
        title = (TextView) view.findViewById(R.id.title);
        type = (TextView) view.findViewById(R.id.type);
        description = (TextView) view.findViewById(R.id.description);
        mFab = (FloatingActionButton) view.findViewById(R.id.fab);


        mFab.setBackgroundTintList(new ColorStateList(new int[][]{new int[0]}, new int[]{0xffffcc00}));

        mToolbar.setTitle("");
        initToolbarNav(mToolbar);
        ImageLoader.loadAll(getActivity(), dataBean.getCover().getDetail(), mImgDetail);
        ImageLoader.loadAll(getActivity(), dataBean.getCover().getBlurred(), bgImage);
        title.setText(dataBean.getTitle());
        StringBuilder stringBuilder = new StringBuilder();
        stringBuilder.append("#").append(dataBean.getCategory())
                .append(" ")
                .append(" / ")
                .append(" ")
                .append(DateUtil.formatTime2(dataBean.getDuration()));
        type.setText(stringBuilder.toString());
        description.setText(dataBean.getDescription());

        mFab.setOnClickListener(v -> {
            if (mFab.getVisibility() == View.VISIBLE) {
                FabTransformation.with(mFab).setListener(new FabTransformation.OnTransformListener() {
                    @Override
                    public void onStartTransform() {
                    }

                    @Override
                    public void onEndTransform() {
                        JumpUtils.goToVideoPlayer(getActivity(), mImgDetail, dataBean);
                    }
                }).transformTo(mImgDetail);
            }
        });

        initColor();

    }

    int color = 0xffffcc00;

    private void initColor() {

        Glide.with(getContext()).load(dataBean.getCover().getBlurred()).asBitmap().into(new SimpleTarget<Bitmap>() {
            @Override
            public void onResourceReady(Bitmap resource, GlideAnimation<? super Bitmap> glideAnimation) {

                Palette.from(resource).generate(palette -> {
                    try {
                        color = palette.getDarkMutedSwatch().getRgb();
                    } catch (Exception e) {
                        LogUtil.d(e.getMessage());
                    }
                    collToolBar.setContentScrimColor(color);
                    mFab.setBackgroundTintList(new ColorStateList(new int[][]{new int[0]}, new int[]{color}));

                });

            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        FabTransformation.with(mFab).setListener(new FabTransformation.OnTransformListener() {
            @Override
            public void onStartTransform() {

            }

            @Override
            public void onEndTransform() {
                if (mImgDetail.getVisibility() == View.INVISIBLE) {
                    mImgDetail.setVisibility(View.VISIBLE);
                }
            }
        }).transformFrom(mImgDetail);
    }

}

File: app/src/main/java/org/cuieney/videolife/kotlin/ui/fragment/video/VideoDetailFragment.kt
package org.cuieney.videolife.kotlin.ui.fragment.video

import android.content.res.ColorStateList
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v7.graphics.Palette
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.request.animation.GlideAnimation
import com.bumptech.glide.request.target.SimpleTarget
import com.konifar.fab_transformation.FabTransformation
import org.cuieney.videolife.R
import org.cuieney.videolife.entity.kaiyanBean.DataBean
import org.cuieney.videolife.kotlin.base.BaseBackFragment
import kotlinx.android.synthetic.main.video_home_detail_fragment.*
import org.cuieney.videolife.common.image.ImageLoader
import org.cuieney.videolife.common.utils.DateUtil
import org.cuieney.videolife.common.utils.LogUtil
import org.cuieney.videolife.kotlin.common.utils.JumpUtils

/**
 * Created by cuieney on 2017/6/2.
 */
class VideoDetailFragment(dataBean: DataBean) :BaseBackFragment(){

    lateinit var dataBean:DataBean
    init {
        val bundle = Bundle()
        bundle.putParcelable(ARG_ITEM,dataBean)
        arguments = bundle
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        dataBean = arguments.getParcelable(ARG_ITEM)
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val inflate = inflater!!.inflate(R.layout.video_home_detail_fragment, container, false)
        return inflate
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        initView()
    }

    fun initView(){
        fab.setBackgroundTintList(ColorStateList(arrayOf(IntArray(0)), intArrayOf(0xffffcc00.toInt())))
        toolbar.title = ""

        initToolbarNav(toolbar)
        ImageLoader.loadAll(activity, dataBean.cover.detail, img_detail)
        ImageLoader.loadAll(activity, dataBean.cover.blurred, bg_image)
        title.text = dataBean.title
        val stringBuilder = StringBuilder()
        stringBuilder.append("#").append(dataBean.category)
                .append(" ")
                .append(" / ")
                .append(" ")
                .append(DateUtil.formatTime2(dataBean.duration.toLong()))
        type.text = stringBuilder.toString()
        description.text = dataBean.description

        fab.setOnClickListener({ v ->
            if (fab.getVisibility() == View.VISIBLE) {
                FabTransformation.with(fab).setListener(object : FabTransformation.OnTransformListener {
                    override fun onStartTransform() {}

                    override fun onEndTransform() {
                        JumpUtils.goToVideoPlayer(activity, img_detail, dataBean)
                    }
                }).transformTo(img_detail)
            }
        })

        initColor()

    }


    internal var color = 0xffffcc00.toInt()

    private fun initColor() {

        Glide.with(context).load(dataBean.cover.blurred).asBitmap().into(object : SimpleTarget<Bitmap>() {
            override fun onResourceReady(resource: Bitmap, glideAnimation: GlideAnimation<in Bitmap>) {

                Palette.from(resource).generate { palette ->
                    try {
                        color = palette.darkMutedSwatch!!.rgb
                    } catch (e: Exception) {
//                        LogUtil.d(e.message)
                    }

                    toolbar_layout.setContentScrimColor(color)
                    fab.setBackgroundTintList(ColorStateList(arrayOf(IntArray(0)), intArrayOf(color)))

                }

            }
        })
    }

    override fun onResume() {
        super.onResume()
        FabTransformation.with(fab).setListener(object : FabTransformation.OnTransformListener {
            override fun onStartTransform() {

            }

            override fun onEndTransform() {
                if (img_detail.getVisibility() == View.INVISIBLE) {
                    img_detail.setVisibility(View.VISIBLE)
                }
            }
        }).transformFrom(img_detail)
    }
}